java/arcs/core/common/Id.kt

```diff
@@ -103,7 +103,16 @@ data class ArcId /* internal */ constructor(
     override val root: String,
     override val idTree: List<String>
 ) : Id {
-    override fun toString() = idToString(this)
+
+    private val arcIdString: String by lazy {
+        idToString(this)
+    }
+
+    override fun toString() = arcIdString
+
+    override fun hashCode() = arcIdString.hashCode()
+
+    override fun equals(other: Any?) = (other as? ArcId)?.toString() == arcIdString
```

Comment:
If the Elvis operator fails, will this return a boolean?

Reply:
if other is not `ArcId`, then `(other as? ArcId)` will be `null`, and so `null?.toString()` will be null. So the expression will be `null == arcIdString` which will evaluate to false, since `arcIdString` will never be null.

Reply:
Thanks for that explanation. There's a lot to unpack in suck a short expression.